### ISSUE TITLE
Something wrong with the Riemann tensor?

### ISSUE THREAD

**Issue by bojone** [sympy/sympy#11799]
I use the following code to calculate the Riemann tensor of ds^2 = f(x,y)^2 (dx^2+dy^2)

`from sympy.diffgeom import Manifold, Patch, CoordSystem
from sympy.diffgeom import TensorProduct as TP
from sympy.diffgeom import metric_to_Riemann_components as Riemann
from sympy.diffgeom import metric_to_Christoffel_2nd as Christoffel
from sympy import Symbol, Function, latex, sin
 
n = 2
M = Manifold('M', n)
P = Patch('P', M)
 
coord = CoordSystem('coord', P, ['x%s'%i for i in range(n)])
x = coord.coord_functions()
dx = coord.base_oneforms()

f = Function('f')
g = [[f(x[0],x[1])**2, 0], [0, f(x[0],x[1])**2]]
metric = sum([g[i][j]*TP(dx[i], dx[j]) for i in range(n) for j in range(n)])

R = Riemann(metric)`

it returns

>>> R[0,1,0,1]
-2*Subs(Derivative(f(_#_1(_Dummy_20), x1), _#_1(_Dummy_20), _#_1(_Dummy_20)), (_#_1(_Dummy_20),), (x0,))/f(x0, x1) - 2*Subs(Derivative(f(x0, _#_0(_Dummy_20)), _#_0(_Dummy_20), _#_0(_Dummy_20)), (_#_0(_Dummy_20),), (x1,))/f(x0, x1) + Subs(Derivative(f(_xi_1, x1), _xi_1), (_xi_1,), (x0,))**2/f(x0, x1)**2 + Subs(Derivative(f(x0, _xi_2), _xi_2), (_xi_2,), (x1,))**2/f(x0, x1)**2

we can see that the coefficient of every term is -2, -2, 1, 1, but the answer the textbook gives is -1, -1, 1, 1. What's wrong? my code or sympy itself?

**Comment by jksuom** [comment]
This looks like a bug. The dummies should not be present in the final result. The relevant [piece of code](https://github.com/sympy/sympy/blob/master/sympy/diffgeom/diffgeom.py#L596) says:
```
        # Remove the dummies
        result = d_result.subs(list(zip(d_funcs, base_scalars)))
        result = result.subs(list(zip(coords, self._coord_sys.coord_functions())))
        return result.doit()  # XXX doit for the Subs instances
```
However, this is what results from the substitutions:
```
(Pdb) p result
Subs(Derivative(f(_#_1(_Dummy_29), x1), _#_1(_Dummy_29), _#_1(_Dummy_29)), (_#_1(_Dummy_29),), (x0,))/f(x0, x1) - Subs(Derivative(f(_xi_1, x1), _xi_1), (_xi_1,), (x0,))**2/f(x0, x1)**2
(Pdb) p result.doit()
2*Subs(Derivative(f(_#_1(_Dummy_29), x1), _#_1(_Dummy_29), _#_1(_Dummy_29)), (_#_1(_Dummy_29),), (x0,))/f(x0, x1) - Subs(Derivative(f(_xi_1, x1), _xi_1), (_xi_1,), (x0,))**2/f(x0, x1)**2
```
So, the dummies are not removed, and, what is worse, the wrong coefficient 2 is introduced by doit().

**Comment by bojone** [comment]
@jksuom is there any method or plan to fix it? 
it seems while use a concrete function to replace f(x0,x1), for example, sin(x0)

**Comment by jksuom** [comment]
One could start by checking out an old version of SymPy where differential geometry was introduced to see if this has ever worked. (That would probably be v.0.7.2.)

If it did work, then one can try to find out when it did stop working, and why, by using git bisect.
If it never did work, then the .subs method should be investigated to see why it would fail.

It is also possible that the `__call__` method of BaseVectorField should be revised to avoid such abundant use of substituted dummies.

**Comment by bojone** [comment]
@jksuom so as you quote, replace "return result.doit()" with "return result" is a tmp solution. That generates right answer.

**Comment by jksuom** [comment]
That is true if we accept solutions with unevaluated Subs instances.
 It seems that .doit() was added to remove those, but obviously that also fails in this case. (That should be fixed independently.)

**Comment by bojone** [comment]
@jksuom it seems the problem is from .doit() or Subs. I have split the expression and get function a:
>>> type(a)
<class 'sympy.core.mul.Mul'>
>>> print a
-Subs(Derivative(f(_#_1(_Dummy_20), x1), _#_1(_Dummy_20), _#_1(_Dummy_20)), (_#_1(_Dummy_20),), (x0,))/f(x0, x1)
>>> a.doit()
-2*Subs(Derivative(f(_#_1(_Dummy_20), x1), _#_1(_Dummy_20), _#_1(_Dummy_20)), (_#_1(_Dummy_20),), (x0,))/f(x0, x1)

**Comment by bojone** [comment]
I find the factor 2 is from Derivative of composite function. the following code
```
from sympy import Symbol, Function

f,g = Function('f'), Function('g')
x = Symbol('x')

f(g(x)).diff(g(x), g(x))
```
returns `2*Derivative(f(g(x)), g(x), g(x))`  and 
`f(g(x)).diff(g(x), g(x)).doit()`
returns `4*Derivative(f(g(x)), g(x), g(x))`

**Comment by jksuom** [comment]
> it seems the problem is from .doit() or Subs.

I agree. And the bug in .doit should definitely be fixed. The substitution failures are annoying but do not give manifestly wrong results.

**Comment by bojone** [comment]
I want to know if this bug has been fixed ? thanks.

**Comment by jksuom** [comment]
No, it is still there in the current master.

**Comment by bojone** [comment]
as I say, the bug is from derivative of composite function. it is a very fundamental bug of a CAS. if it is not fixed, how can I trust the final result of a complex function of sympy
